docs/dependency-injection.md

```diff
@@ -0,0 +1,190 @@
+
+Blazor's DI system is responsible for connecting component abstractions with their concrete implementations. To connect the abstractions with the implementations, DI is configured during startup of the app. An example is shown later in this topic.
+
+## Use of existing .NET mechanisms
+
+DI in Blazor is based on .NET's [System.IServiceProvider](https://docs.microsoft.com/dotnet/api/system.iserviceprovider) interface. It defines a generic mechanism for retrieving a service object in .NET apps.
+
+Blazor's implementation of `System.IServiceProvider` obtains its services from an underlying [Microsoft.Extensions.DependencyInjection.IServiceCollection](https://docs.microsoft.com/dotnet/api/microsoft.extensions.dependencyinjection.iservicecollection).
+
+## Add services to DI
+
+After creating a new Blazor app, examine the `Main` method in *Program.cs*:
+
+```csharp
+static void Main(string[] args)
+{
+    var serviceProvider = new BrowserServiceProvider(configure =>
+    {
+        // Add any custom services here
+    });
+
+    new BrowserRenderer(serviceProvider).AddComponent<App>("app");
+}
+```
+
+`BrowserServiceProvider` receives an action with which you can add your app services to Blazor's DI. `configure` is referencing the underlying `IServiceCollection`, which is a list of service descriptor objects ([Microsoft.Extensions.DependencyInjection.ServiceDescriptor](https://docs.microsoft.com/dotnet/api/microsoft.extensions.dependencyinjection.servicedescriptor)). Services are added by providing service descriptors to the service collection. The following is a code sample demonstrating the concept:
+
+```csharp
+static void Main(string[] args)
+{
+    var serviceProvider = new BrowserServiceProvider(configure =>
+    {
+        configure.Add(ServiceDescriptor.Singleton<IDataAccess, DataAccess>());
+    });
+
+    new BrowserRenderer(serviceProvider).AddComponent<App>("app");
+}
+```
+
+`ServiceDescriptor` offers various overloads of three different functions that can be used to add services to Blazor's DI:
+
+| Method      | Description |
+| ----------- | ----------- |
+| `Singleton` | Blazor creates a *single instance* of your service. All components requiring this service receive a reference to this instance. |
+| `Transient` | Whenever a component requires this service, it receives a *new instance* of the service. |
+| `Scoped`    | Blazor doesn't currently have the concept of DI scopes. `Scoped` behaves like `Singleton`. Therefore, prefer `Singleton` and avoid `Scoped`. |
+
+## System services
+
+Blazor provides default services that are automatically added to the service collection of a Blazor app. The following table shows a list of the default services currently provided by Blazor.
```

Comment:
I wonder if it's worth clarifying that these are the default services in a `BrowserServiceProvider` (which is what the project template uses by default in `Program.cs`). If someone creates their own different service provider, it wouldn't contain these services.